### PR TITLE
[prover] Add support of using receiver style `self` in struct invariants

### PR DESCRIPTION
**Pull request by rahxephon89**
## Description

This PR allows users to use the receiver style `self` in struct invariants when the language version is at least 2.0. For instance, for the struct below:

```
module 0x42::TestInvariants {
    struct PositionalStruct(u64, u64) has key, copy, drop;
}
```

We can write the invariant as below:

```
   spec PositionalStruct {
        invariant self.0 > 10;
        invariant self.1 < 10;
    }
```

Note that if a struct has a field called `self`:

```
    struct StructSelf has key, copy, drop {
        self: u64
    }
```

We need to use `self.self` to access the field instead of directly using `self`:

```
    spec StructSelf {
        invariant self.self > 10;
    }
```

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Existing tests pass and new tests.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by rahxephon89** [comment]
* **#14265** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14265?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rahxephon89 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: third_party/move/move-model/src/builder/module_builder.rs [hunk 1]
```diff
@@ -1777,6 +1777,12 @@ impl<'env, 'translator> ModuleBuilder<'env, 'translator> {
                 if let StructLayout::Singleton(fields, _is_positional) = &entry.layout {
                     et.enter_scope();
                     for f in fields.values() {
+                        if et.env().language_version.is_at_least(LanguageVersion::V2_0)
```
Comment (brmataptos) [inline third_party/move/move-model/src/builder/module_builder.rs:1780]: Can you pull this language version lookup out of the loop?  I don't think Rust will be able to CSE it.  It also could be clearer to give it a name.

Comment (brmataptos) [reply third_party/move/move-model/src/builder/module_builder.rs:1780]: Also, can you add a brief comment to explain what is happening here?  IIUC,
```
// In Aptos Move 2.0 and above, field `self` is omitted from local bindings so `self` can be used to refer to `self` parameter.
```

Comment (rahxephon89) [reply third_party/move/move-model/src/builder/module_builder.rs]: done

File: third_party/move/move-model/src/builder/module_builder.rs [hunk 2]
```diff
@@ -1789,6 +1799,12 @@ impl<'env, 'translator> ModuleBuilder<'env, 'translator> {
                             None,
                         );
                     }
+                    if et.env().language_version.is_at_least(LanguageVersion::V2_0) {
```
Comment (brmataptos) [inline third_party/move/move-model/src/builder/module_builder.rs:1802]: use `lang_ver_ge_2` here, too.

**Review by brmataptos — APPROVED** [review]

File: third_party/move/move-model/src/builder/module_builder.rs (hunk 2 above)
Comment (rahxephon89) [reply third_party/move/move-model/src/builder/module_builder.rs]: done

**Review by wrwg — APPROVED** [review]

**Pull request merged**